cf-custom-resources/lib/cert-replicator.js

```diff
@@ -0,0 +1,320 @@
+    });
+
+    const parsedUrl = new URL(event.ResponseURL || defaultResponseURL);
+    const options = {
+      hostname: parsedUrl.hostname,
+      port: 443,
+      path: parsedUrl.pathname + parsedUrl.search,
+      method: "PUT",
+      headers: {
+        "Content-Type": "",
+        "Content-Length": responseBody.length,
+      },
+    };
+
+    https
+      .request(options)
+      .on("error", reject)
+      .on("response", (res) => {
+        res.resume();
+        if (res.statusCode >= 400) {
+          reject(new Error(`Error ${res.statusCode}: ${res.statusMessage}`));
+        } else {
+          resolve();
+        }
+      })
+      .end(responseBody, "utf8");
+  });
+};
+
+/**
+ * Replicate a public certificate from AWS Certificate Manager in the target region.
+ *
+ * @param {string} requestId the CloudFormation request ID
+ * @param {string} appName the application name
+ * @param {string} envName the environment name
+ * @param {string} certArn arn for the certificate to replicate from
+ * @param {string} envRegionAcm acm client in environment region
+ * @param {string} targetRegionAcm acm client in target region
+ * @returns {string} ARN of the replicated certificate
+ */
+const replicateCertificate = async function (
+  requestId,
+  appName,
+  envName,
+  certArn,
+  envRegionAcm,
+  targetRegionAcm
+) {
+  const { Certificate } = await envRegionAcm
+    .describeCertificate({
+      CertificateArn: certArn,
+    })
+    .promise();
+  const domainName = Certificate.DomainName;
+  const sans = Certificate.SubjectAlternativeNames;
+
+  const crypto = require("crypto");
+  return targetRegionAcm
+    .requestCertificate({
+      DomainName: domainName,
+      SubjectAlternativeNames: sans,
+      IdempotencyToken: crypto
+        .createHash("sha256")
+        .update(requestId)
+        .digest("hex")
+        .substr(0, 32),
+      ValidationMethod: "DNS",
+      Tags: [
+        {
+          Key: "copilot-application",
+          Value: appName,
+        },
+        {
+          Key: "copilot-environment",
+          Value: envName,
+        },
+      ],
+    })
+    .promise();
+};
+
+/**
+ * Deletes a certificate from AWS Certificate Manager (ACM) by its ARN.
+ * Specifically, if it is the last certificate attaching to the listener, it will also remove the CNAME records
+ * for validation in all the root, app, and env hosted zones in parallel.
+ * If the certificate does not exist, the function will return normally.
+ *
+ * @param {string} arn The certificate ARN
+ * @param {string} targetRegionAcm acm client in target region
+ */
+const deleteCertificate = async function (arn, acm) {
+  try {
+    let inUseByResources;
+
+    for (let attempt = 0; attempt < maxAttempts; attempt++) {
+      const { Certificate } = await acm
+        .describeCertificate({
+          CertificateArn: arn,
+        })
+        .promise();
+
+      inUseByResources = Certificate.InUseBy || [];
+      if (!inUseByResources.length) {
```

Comment:
nit: just to make it clear
```suggestion
      if (inUseByResources.length === 0) {
```